opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -212,7 +214,19 @@ namespace Opm {
             }
             // Apply as Schur complement the well residual to reservoir residuals:
             // r = r - duneC_^T * invDuneD_ * resWell_
-            well->apply(res);
+            // Well equations B and C uses only the perforated cells, so need to apply on local residual
+            auto cells = well->cells();
```

Comment:
Use `const auto&` here, current code is making a copy. Also on lines 256, 1575 and 1620. And 1581, and 1627 in a slightly different context. The copies made on 1575 and 1620 are potentially overwritten, if in an NLDD context, so it is not straightforward to make those const references, but require some refactoring.

Reply:
Made all references to the well cells, except the one for NLDD, where it is a reference to the pre-calculated values, that is updated at each timestep.